cache/client.go

```diff
@@ -0,0 +1,122 @@
+package cache
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/merkletree"
+)
+
+func verifyBlockHeader(header types.BlockHeader, minTimestamp types.Timestamp) error {
+	// Check that the timestamp is not too far in the past to be acceptable.
+	if minTimestamp > header.Timestamp {
+		return fmt.Errorf("Block header validation failed: EarlyTimestamp")
+	}
+
+	// Check if the block is in the extreme future. We make a distinction between
+	// future and extreme future because there is an assumption that by the time
+	// the extreme future arrives, this block will no longer be a part of the
+	// longest fork because it will have been ignored by all of the miners.
+	if header.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
+		return fmt.Errorf("Block header validation failed: ExtremeFutureTimestamp")
+	}
+
+	// Check if the block is in the near future, but too far to be acceptable.
+	// This is the last check because it's an expensive check, and not worth
+	// performing if the earlier checks failed.
+	if header.Timestamp > types.CurrentTimestamp()+types.FutureThreshold {
+		return fmt.Errorf("Block header validation failed: FutureTimestamp")
+	}
+	return nil
+}
+
+// minimumValidChildTimestamp returns the earliest timestamp that a child node
+// can have while still being valid.
+func minimumValidChildTimestamp(headers []types.BlockHeader, headerIndex int) (minTimestamp types.Timestamp, err error) {
+	// Get the previous MedianTimestampWindow timestamps.
+	windowTimes := make(types.TimestampSlice, types.MedianTimestampWindow)
+	windowTimes[0] = headers[headerIndex].Timestamp
+	parent := headers[headerIndex].ParentID
+	for i := 1; i < int(types.MedianTimestampWindow); i++ {
+		// If the genesis block is 'parent', use the genesis block timestamp
+		// for all remaining times.
```

Comment:
Maybe we can move this case outside of the loop and just return genesis block timestamp early?

Reply:
This isn't just for genesis case itself, but for all the blocks which have genesis block less than types.MedianTimestampWindow times behind. This is the computation rule for them.

Reply:
```
if number of blocks < types.MedianTimestampWindow {
  return timestamp of genesis block
}
loop
```

Reply:
The algorithm here is different, we add the actual timestamps of blocks till genesis block, after that we fill the remaining spots (if any) with GenesisTimestamp till `MedianTimestampWindow` length. We finally return the median of this list, which is not necessary equal to the timestamp of genesis block.

Reply:
I see.